lang/semgrep-grammars/src/semgrep-sml/grammar.js

```diff
@@ -0,0 +1,46 @@
+/*
+  semgrep-sml
+
+  Extends the standard sml grammar with semgrep pattern constructs.
+*/
+
+const base_grammar = require('tree-sitter-sml/grammar');
+
+module.exports = grammar(base_grammar, {
+  name: 'sml',
+
+  conflicts: ($, previous) => previous.concat([
+    [$.mrule, $._atpat]
+  ]),
+
+  /*
+     Support for semgrep ellipsis ('...') and metavariables ('$FOO'),
+     if they're not already part of the base grammar.
+  */
+  rules: {
+    semgrep_ellipsis: $ => '...',
+
+    // As it turns out, $x parses as the application of `$` to an identifier.
```

Comment:
yes but how is parsed $ X
with a space in between?
Probably better to make a special token for $XXX with token.immediate. There should be no conflict.